TEST-#3422: always import read data to OmniSci in IO tests

dchigarev opened the pull request:
Signed-off-by: Dmitry Chigarev <dmitry.chigarev@intel.com>

<!--
Thank you for your contribution!
Please review the contributing docs: https://modin.readthedocs.io/en/latest/CONTRIBUTING.html
if you have questions about contributing.
-->

## What do these changes do?

This PR adds explicit import of the read data to OmniSci at every IO test. Modification of the `eval_io` function was used where possible, in other places it was done by the `ForceOmnisciImport` context class. Both of these were placed at the newly created utility file `experimental/engines/omnisci_on_native/test/utils.py`, other testing helper functions that previously were located at the file with tests were also moved to the `utils.py`.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #3422 <!-- issue must be created for each patch -->
- [x] tests <s>added and</s> passing
- [x] module layout described at `docs/developer/architecture.rst` is up-to-date <!-- if you have added, renamed or removed files or directories please update the documentation accordingly -->

dchigarev commented on modin/pandas/test/utils.py:
```diff
@@ -773,6 +773,7 @@ def call_eval_general():
             pd,
             pandas,
             applyier,
+            comparator=comparator,
```
it appeared that this parameter hadn't been forwarding

dchigarev commented on modin/experimental/engines/omnisci_on_ray/test/utils.py:
```diff
@@ -0,0 +1,173 @@
+    )
+
+
+def trigger_import(*dfs):
```
this is almost a full copy of the [`asv_bench.benchmarks.utils.common.trigger_import`](https://github.com/modin-project/modin/blob/0a642753af2cd2332ac71166bcddc89b0cace9d5/asv_bench/benchmarks/utils/common.py#L443-L468) function. We can't import that function from `asv_bench`, since ASV stuff is not a part of the modin package, we can't import this function to `asv_bench` from here either, because this will break ASV compatibility with the previous version of modin.

dchigarev commented on modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:
```diff
@@ -186,11 +109,13 @@ def test_housing_csv(self):
-            rm = to_pandas(pd.read_csv(csv_file, engine="arrow", **kwargs))
-            assert rp is not None
-            assert rm is not None
-            # TODO: df_equals(rp, rm)  #  needs inexact comparison
```
inexact comparison appeared to be unneeded, output frames are identical

ienkovich commented on modin/experimental/engines/omnisci_on_ray/test/test_dataframe.py:
```diff
@@ -251,6 +188,7 @@ def test_null_col(self, null_dtype):
             dtype={"a": "int64", "b": "int64", "c": null_dtype},
             skiprows=1,
         )
+        trigger_import(exp)
```
Due to the following computations, we will have execution and therefore import triggered anyway for `exp`.

dchigarev replied:
moved import triggering to the place before casting to pandas and also added a check to do not import the same frame multiple times

dchigarev commented:
@modin-project/modin-omnisci @modin-project/modin-core just a reminder to review

amyskov commented:
Looks great! A little bit confusing that imported to OmniSci table doesn't used during pandas/Modin results comparison by the `comparator`, shouldn't it be exported back for checking?

ienkovich commented:
> Looks great! A little bit confusing that imported to OmniSci table doesn't used during pandas/Modin results comparison by the `comparator`, shouldn't it be exported back for checking?

We keep arrow tables for imported partitions and therefore never need to export them back.

dchigarev commented:
After adding modified `eval_io` to the `pandas/test/test_io.py` with ~500 `read_csv` tests it appeared that the `test_io.py::TestCsv` step is failing with an out-of-memory error. I've added a print statement that prints the amount of used system RAM in tests, the job is being killed shortly after the RAM usage [hits 100%](https://github.com/modin-project/modin/pull/3423/checks?check_run_id=3633094484#step:8:1384).

The reason for hitting 100% RAM usage is that this step has a lot of `read_csv` calls, which generates a lot of tables in OmniSci, which are never going to be deleted because there is no mechanism of garbage collecting of orphaned tables.

@ienkovich is this awared problem of no garbage collection of unused tables in OmniSci and is there a way to solve this problem in the context of this PR (something like cycle-rewriting or similar)?

dchigarev commented:
In order to resolve RAM overflow by Omnisci tables I've added a mechanism of dropping them at the end of every test that requires forcibly import read data to Omnisci (done by [`ForceOmnisciImport`](https://github.com/dchigarev/modin/blob/40fd7d71f57341bdabf5120ee77db5431c32cdb6/modin/experimental/engines/omnisci_on_native/test/utils.py#L69) context-class). 

After that, it revealed that one of the IO tests consumes ~17Gb of RAM when data is imported to Omnisci (mentioned test is [`test_read_csv_default_to_pandas_url`](https://github.com/dchigarev/modin/blob/40fd7d71f57341bdabf5120ee77db5431c32cdb6/modin/pandas/test/test_io.py#L971)). The reason for that is that in this test it reads a many-column frame (~1000 columns) where each column contain only strings. It seems that it's really unoptimized case for Omnisci, so I've added skipping this test if available RAM is less than 17Gb and selected backend is Omnisci.

dchigarev commented:
@amyskov @ienkovich @modin-project/modin-core PR is ready for review

anmyachev commented on modin/pandas/test/test_io.py:
```diff
@@ -952,6 +957,13 @@ def test_read_csv_default_to_pandas(self):
+            Backend.get() == "Omnisci"
+            and psutil.virtual_memory().available < 17 * 2 ** 30
+        ),
+        reason="It takes about ~17Gb of RAM for Omnisci to import table from this test because of too many (~1000) string columns in it.",
```
We can use subset of all columns.

dchigarev replied:
done

vnlitvinov commented on modin/experimental/engines/omnisci_on_native/test/utils.py:
```diff
@@ -0,0 +1,195 @@
+            self._imported_frames.append((df, frame_id))
+
+    def __enter__(self, *dfs):
+        pass
```
it should be returning `self` so things like
```python
with ForceOmnisciImport(df) as foo:
    print(foo)
```
work correctly. Not that it is required currently, but it's better to implement a general context manager protocol from the get go.

dchigarev replied:
done

vnlitvinov commented on modin/experimental/engines/omnisci_on_native/test/utils.py:
```diff
@@ -0,0 +1,195 @@
+        self._imported_frames = []
+
+
+def set_execution_mode(frame, mode, recursive=False):
```
can we have a small docstring here, at least describing what this does and why it's needed?

dchigarev replied:
added

vnlitvinov commented on modin/experimental/engines/omnisci_on_native/test/utils.py:
```diff
@@ -0,0 +1,195 @@
+                constructor_kwargs=constructor_kwargs,
+                **kwargs,
+            )
+            _ = exp_res.index
```
why is this needed?

dchigarev replied:
it seems that it was done in order to trigger actual computations (to get the index axis we need the frame to be computed)

BTW, I haven't changed this function as well as `set_execution_mode`, these were just copied here from the `test_dataframe.py`

vnlitvinov reviewed (approved):
LGTM, thanks!

YarShev commented on modin/experimental/engines/omnisci_on_native/test/utils.py:
```diff
@@ -0,0 +1,216 @@
+            if df.empty:
+                continue
+            partition = df._query_compiler._modin_frame._partitions[0][0]
+            if partition.frame_id is not None:
```
Shouldn't this be like `if partition.frame_id is None:`?

dchigarev replied:
`frame_id` holds the table name that is imported to OmniSci, which corresponds to the `partition` DataFrame. If `frame_id` is None, then no import has been triggered yet, so we have to stay in this for loop and import the table, if `frame_id` is not None, then the table has been already imported and we have to skip this frame.

YarShev replied:
I got, thanks!
